### DISCUSSION TITLE
RFC: Next.js <Link to=""> to handle dynamic routes

### DISCUSSION THREAD

**Discussion by Timer** [vercel/next.js#8207] [RFC]
# Feature request

## Is your feature request related to a problem? Please describe.

Next.js should be able to deduce what route should be rendered for a given URL.

Right now, Next.js requires you specify both the original page and route:
```jsx
<Link href="/news/[id]" as="/news/1234">
  <a> ... </a>
</Link>

// or

<Link href="/news/[id]?sort=desc" as="/news/1234?sort=desc">
  <a> ... </a>
</Link>
```

However, Next.js should know that `/news/1234` matches `/news/[id]` as it does with the production web server.

## Describe the solution you'd like

Next.js should accept a Link prop that automatically handles this behavior, potentially like so:
```jsx
<Link to="/news/1234"> {/* knows to render `./pages/news/[id].js` */}
  <a>News Article 1234</a>
</Link>

// or

<Link to="/news/1234?sort=desc"> {/* knows to render `./pages/news/[id].js` */}
  <a>News Article 1234</a>
</Link>
```

This proposal uses `to` instead of making `href` have different behavior so we could potentially make it opt-into not using the child component, instead, rendering `<a>` automatically like React Router or Reach Router do.

Meaning:
```jsx
<Link to="/news/1234"> {/* knows to render `./pages/news/[id].js` */}
  News Article 1234
</Link>
```

## Describe alternatives you've considered

Next.js could fall back to matching a dynamic route if not found:
```jsx
<Link href="/news/1234"> {/* knows to render `./pages/news/[id].js` */}
  <a> ... </a>
</Link>
```

## Additional context

Next.js currently does not ship a routing manifest, and that'd be a requirement for this feature to work.
This routing manifest would make the app size heavier than if the user always instructed Next.js with `href` and `as`, but some could argue the DX is worth it.

## TBD

1. We should ban trailing slashes in `to` so we can make this a framework feature, i.e. `trailingSlash: true`.

**Comment by flippidippi** [comment]
It would also be helpful here if query strings behave correctly like `to="/news/1234?topic=5678"`, where the query would be populated with `{ id: 1234, topic: 5678 }`.

When the page is known an improvement to the API for pathname/query like `{ pathname: "news/[id]", query: { id: 1234, topic: 5678 } }`, which would send you to `news/1234?topic=5678`. Not sure if this would fall under using `to=` but I think it would be super helpful.

**Comment by martpie** [comment]
Could there be a manifest only server-side? And when a Link is clicked, the server answers with which bundle/page to render?

Probably slower, but we avoid shipping a manifest that may not scale.

**Comment by Timer** [comment]
@martpie if we explored the manifest route (instead of static analysis), we'd probably make the manifest be downloaded asynchronously and not block application hydration. 

Routing events before that manifest arrived would revert to server-side rendering IMO, not blocking while fetching the manifest.

**Comment by vitorbal** [comment]
_(x-posting from slack)_

Hey all :wave: found this issue while doing some research and wanted to chime in re: routing manifest.

I’m sharing this in the hopes it’s relevant to help decide if we want a routing manifest or not:

At my company, we’re transitioning to Next.js page by page, which means some Next.js pages need to link to non-Nextjs pages.

Whenever a page gets migrated to Next.js, having to go back and wrap any <a> tags that link to that page with a `<Link>` would be a hassle.

If we could build a component that can decide if it needs to wrap an <a> tag in a `<Link>` component or not based on a routing manifest, migrating pages to Nextjs would be easier:

Having that routing manifest gives us a way to check if an url matches any routes, sort of like this:

```
const MyLink = props => {
  const { text, href } = props;
  const isNextRoute = routes.indexOf(href) !== -1;

  if (isNextRoute) {
    return (
      <Link href={href} /* ... */>
        <a>{text}</a>
      </Link>
    );
  }

  return <a href={href}>{text}</a>;
};
```


We also have a shared design system that needs to be rendered by both Next.js and non-Next.js pages – meaning, if we have e.g. a navigation component, that navigation component can’t render `<Link>`s for all the nav items because it doesn’t know if those URLs are pointing to next.js pages, non-next pages, or even external pages.

This makes the API for this component more complex, because expecting an array of URLs doesn’t suffice:

```
<Nav items={[
  { text: 'Overview', href: '/overview' },
  { text: 'Integrations', href: '/integrations' },
  { text: 'Updates', href: '/updates' }
]} />
```

^ `<Nav` doesn’t know if it’s okay to wrap the anchor tags in `<Link/>` components or not.

Again, if there would be a way to check if an url matches any routes, we could keep the API of the `Nav` component lean by injecting our `MyLink` component into our `Nav` component (or into our design system via a context provider):

```
<Nav
  LinkComponent={MyLink}
  items={[
    { name: 'Overview', href: '/overview' },
    { name: 'Integrations', href: '/integrations' },
    { name: 'Updates', href: '/updates' },
  ]}
/>
```

Hopefully this is useful input 🙂

**Comment by jstcki** [comment]
I'd like to add a suggestion which would work without static analysis or a manifest or even a separate `to` prop, based on a solution I have implemented in an app:

It could be as simple as making the `Link` component smart enough to replace dynamic route parts in `pathname` automatically based on the `query` when using the url-object form of `href`:

```js
<Link href={{ pathname: "/foo/[bar]", query: { bar: 1, baz: 2 } }}><a>...</a></Link>
```

Currently this renders `<a href="/foo/[bar]?bar=1&baz=2">...</a>` which is _functional_ but odd. So I have to replace the dynamic parts in `pathname` with matching `query` params and stringify a `search` to construct an `as` prop manually.

If this was handled out of the box, the rendered link would be `<a href="/foo/1?baz=2">...</a>`. Which is what I'd expect.

Another example to illustrate the issue where it's quite tedious to construct the `as` prop would be to replace a query param of the current route to link to another locale:

```js
const LinkCurrentRouteToLocale = ({ locale, ...rest }) => {
  const { pathname, query } = useRouter();
  // pathname being /[locale]/foo or similar
  return <Link href={{ pathname, query: { ...query, locale } }} {...rest} />
}
```

**Comment by pepf** [comment]
@herrstucki Not sure if this is exactly this issue, but I'm running into exactly this scenario as well (mixing up regular query params with the "dynamic" ones). 

This becomes especially cumbersome if you'd have a page with one or more "dynamic route" params which has to support one or more variable query params (think in the direction of utm tags).  The query params just seem to "blow up" at some point 😅

**Comment by Ephem** [comment]
I'd love this! Building a PoC for moving over to Next and use the filesystem based routing we very quickly ran into this problem since we have a lot of dynamic/unknown links from the backend which would be hard to change. We do have a custom, hacky, userland solution based on some Next-internals in the works..

Where are you leaning implementation-wise? As I see it, static analysis seems hard for dynamic links coming from APIs, and while not the case for us, shipping the entire manifest might be problematic for very large applications, even if it's lazy?

I haven't thought this through all the way, but could one approach be to track all the `to` that were used at runtime? On the server, match and extract only those parts of the manifest and inject in payload (or lazily). On the client, since links can change at any time, batch discovery of new `to` and request manifest for that lazily, fallback to full reload if not available, possibly after waiting a short time for it to come in if request is in flight?

Just out of curiosity, is this something you would be interested in external help on or is it mainly something you need to incubate on in relation to the bigger picture? 😄

**Comment by Ephem** [comment]
Hm, I just realized that approach would only work with ssr, not static so that's a bummer. Pretty complex solution too compared to just shipping a full manifest lazily which should probably be fine for the vast majority of cases, and you can still optimize that away by using `href`+`as`.

**Comment by jeysal** [comment]
I "solved" this problem in a closed-source project, I think in a somewhat similar way to what Next would do to implement it by itself, just not really as nice and polished. This is what we did:

---

In the app's `getInitialProps`, on the server, get all routes (`pagePaths`) from the `nextApp.pagesManifest` (we have access to the `nextApp` instance by setting it onto our Express `app` instance).
In dev, fall back to no routes because there's no comprehensive list of all routes available. This means that there will be full page reloads in dev, but we accept that as a kind of omitted performance optimization, of which there are many others in dev vs prod.
In the app's `getInitialProps` on the client, reuse the `pagePaths` the server initially generated. They would only change on redeploy anyway and stale ones aren't a critical issue, so not worth to refetch from the server all the time.
```js
export const getPagePaths = req => (req
  ? getSortedRoutes(Object.keys(req.app.locals.nextApp.pagesManifest || {/* in dev */}))
  : window.__NEXT_DATA__.props.pagePaths);
```

---

We then pass them into a React context that is consumed by our custom link component. That component gets an anchor tag as a child, the corresponding `href` as a prop, and our `pagePaths` from the context. It then uses
```js
import {
  getRouteMatcher, getRouteRegex, getSortedRoutes, isDynamicRoute,
} from 'next/dist/next-server/lib/router/utils';
```
to decide whether to
* render a plain anchor, for unknown/external links
* render a Next Link with `href`, for links to static pages, or
* render a Next Link with `href` and `as`, for links to dynamic pages.
```js
({ children, href, pagePaths }) => {
  const matchingPagePath = pagePaths.find(
    path => getRouteMatcher(getRouteRegex(path))(href),
  );
  if (matchingPagePath) {
    if (isDynamicRoute(matchingPagePath)) {
      // Next Link requires href and as for dynamic routes
      return <Link href={matchingPagePath} as={href}>{children}</Link>;
    }
    return <Link href={href}>{children}</Link>;
  }

  return children;
}
```

---

I hope it helps those of you who need a comprehensive solution that does not require doing extra work everywhere `Link` is used. Obviously proceed with caution and evaluate how urgently you need this as there's imports from private paths etc etc

**Comment by Timer** [comment]
Please be aware that importing from `dist` may break in any release (even patch), and is covered by absolutely zero semver guarantees. I strongly advise against this.

**Comment by jeysal** [comment]
Yes, thanks @Timer, I mentioned that in the final paragraph 😄

**Comment by jayu** [comment]
Would it be possible to ship a solution similar to what @jeysal code is doing? I mean is it OK with library assumptions? 
I could help with the implementation if a solution based on `pagesManifest` send to the client and matching inside `<Link/>` component is an acceptable way to achieve to goal

**Comment by Timer** [comment]
@jayu the solution posted by @jeysal should be avoided in any production facing app. It **will break** without warning.

We want to solve this first-class in Next.js soon, however, there's no ETA.

**Comment by jeysal** [comment]
I just want to relativize that statement a bit and explain why we went for it anyway to help others make the judgement whether it may be worth it for them since this appears to be a quite popular issue.
It is very easy to take the dogmatic approach and declare private imports as the greatest of evils, I know this very well as an OSS maintainer myself.

I believe, however, that we need to look at each case individually, and I would classify this particular instance as a relatively clean solution that relies on existing concepts in Next instead of implementing everything in a potentially diverging way and is relatively low-risk.
First of all, let me clarify that **you need to have your Next version locked down** for this. Otherwise, yes, your master branch will break suddenly and it may come at a very inconvenient time where you may e.g. be under time pressure busy with fixing other things.

That said, I see two kinds of breaking changes that are likely to occur on the internal APIs we're using (others may be possible but seem unlikely):
* changing the behavior of one of the imports and
* changing the internal package structure with an effect on one of the imports.

As an example of the former, let's imagine Next.js introduces a new way to declare dynamic routes using curly braces in a `/{id}/edit.jsx` pattern.
In a case like this, the behavior of at least `isDynamicRoute` and `getSortedRoutes` will change, but I would say it is very _likely_ that we will (like in the example given) also want this new behavior in our usage of those functions - after all the semantics we're interested in for this solution on a conceptual level is checking whether a *route is dynamic*, and as the export says in its name, that is what it does.

The latter may mean (re)moving the file being imported from, renaming the exports, or removing the logic entirely in a refactor of the whole routing mechanics.
In this case, when we try to upgrade from our locked-down version of Next to the new "breaking" version, we will notice the breakage quickly and track down its location easily.
We will then need to invest development time (to a variable extent, based on the nature of the internal changes in Next) into upgrading to that version of Next; I do not deny that this could be a problem and is a factor to consider.

However, I believe that most real world applications have technical debt and architecture choices posing a way greater risk to the continued stability and maintainability of the software than this workaround.  Like those, the risk posed by this workaround can also be absorbed to a degree by integration/e2e tests.
For this reason, _I deny the existence of a strong correlation between the production readiness of the software and its usage of private imports **in general**_.

For us, the risk posed by this workaround appeared to be lower than that posed by the choice to introduce additional work for every Link we add, especially because we have abstractions built on top of `Link` that cause the `href` and `as` to not be colocated with the actual usage of `Link` in our code.

Anyway, I think I was in a bit of a philosophical mood and this became more of a general braindump of my thoughts on this topic than directly related to this issue.
If I wrote any more I may as well just spend the time on actually implementing the feature ;D
Apologies for making this issue thread longer than it has to be, but maybe this will be thought-provoking if someone actually reads through it 😅
@Timer don't feel pressured to respond with a wall of text, this is not an attack, I just wanted to write down some thoughts I had on this topic. As stated in the beginning I see where you are coming from :)
When this issue is resolved, naturally we intend to migrate to the official solution.

**Comment by timneutkens** [comment]
We've seen many times before that people tried importing internals, in basically all cases this resulted in:
- An issue posted on Next.js saying "Why did you break this stuff now I have to refactor" "Why are you making breaking changes" and similar replies.
- Lost trust from managers in "upgrading Next.js to improve results" as it would "break" which is actually caused by the engineers that worked on the project, not by us introducing breaking changes.

Hence why we recommend not importing internals like @timer was saying.

**Comment by jeysal** [comment]
Of course, I'm not asking you to endorse it, not would I encourage people to complain about breaking changes that aren't actually breaking changes because they only affect internals.
I just think indiscriminately relating it to production readiness is a bit far-fetched - it can be the right choice :)

**Comment by jayu** [comment]
I think you get me wrong :D The intention of my post was to confirm if the solution based on sending page pages manifest (or whatever is needed to handle route matching on client-side) via data assigned to a global variable is acceptable. Then it would be possible to perform matching inside `<Link />` component, preferably using the logic that is used on the server. I understand that using the next builtins is a bad idea and that's why I would like to contribute and implement it *inside* Next. However, I didn't look into the source code yet, so maybe this approach will require sharing code between Next packages or something.
I'm asking to avoid spending my time on the contribution that might not be accepted because it is bad by definition. For some reason, you didn't implement this yet.  I'm wondering if there is any other reason than priorities and lack of time, especially if it seems that the logic to solve this problem has been already implemented. Maybe it is insecure to share pages information with the client-side but it seems to be already sent. 
I can spend time on creating workarounds that will work in my case and have to maintain them later, or I can use the same time to contribute and solve the problem at the source and make people lives easier :) 
I will be migrating a quite big project to Next and I will have to implement it anyway.

**Comment by jeysal** [comment]
@jayu I can only speak for myself, but I understood the intention of your post - but regarding the implementation I'd say there's definitely a lot missing for an actual proper implementation inside Next. I probably noticed more problems while working on this that I've already forgotten now, but to name two example things that need additional work to convert the workaround into a real solution:
* As you mentioned, it makes the client use server routing logic. This would at the very least require extraction to shared utils.
* It only works in prod mode, because in dev `pagesManifest` is `undefined`. Even if we used another source available in dev to retrieve a list of pages, let's say the pages manifest JSON file in the next build dir, that does not contain all routes in existence either, but only those that have been accessed via the dev server recently. This is because afaik the Next dev server in fact *does not know* (as in, does not hold a list in memory of) which routes exist. Instead, it hits the FS when you request a page and periodically forgets about them; you can see this happening in the console. I imagine this is for performance reasons in apps with lots of pages.

**Comment by BramDecuypere** [comment]
Will this also include an optimisation for when exporting an URL?

```
<Link href="/overview" as="/nl/leasing-auto" >
  Search vehicles
</Link>
```

Results in an URL like this: instead of with the `as` route?
```
<a href="/overview">
  Search vehicles
</a>
```
Expected:
```
<a href="/nl/leasing-auto">
  Search vehicles
</a>
```

Or is it possible to achieve this in some other way?

**Comment by keverw** [comment]
As someone who's been learning more of the modern JS, found it interesting links are handled like thagt. Matching how React Router's own <link> element like mentioned sounds prefect. 

I'm not that far yet in my project, but one of the ideas is a knowledge base and blog. Which could link out anywhere, but as long as it' the same app(subdomain, etc...) my plan was to change those to `<link>` so having to do things like `/news/[id]` sounds like a barrier to that idea. Maybe early on could manage, but as more links or if you grow to have a marketing team I doubt many of them have the concept of url routers, but copying and pasting a url is like second nature to them.

Also another pro for matching how React does it, say you make a component you reused between Next and Create React for say different products, could set the `<link>` for either toolset your using - maybe as a context to help with code reuse.

**Comment by saibs** [comment]
For a better DX, Is it possible to have typesafe routing for dynamic routes something in a direction to support at IDE level? 🤔

**Comment by mikestopcontinues** [comment]
I hacked it by reading the `/pages` directory instead. I'm doing this with `babel-plugin-preval`, but you could easily do it in a script prior to running `next`, then import the json results.

Next could totally generate a route map as seen below, and in dev mode, replace it. 

`./prevalRoutes.js`: Notice, I generate regex strings here and map them to RegExp types in my component since `preval` can only pipe json-content.
```js
const path = require('path');
const globby = require('globby');
const R = require('ramda');

// this could be compressed into one reduce function. was just a first pass
module.exports = R.pipe(
  // get the files
  () => globby.sync('**/*.js', {
    cwd: path.resolve(process.cwd(), 'pages'),
    ignore: ['api', '_*.js'],
  }),
  // clean em up
  R.map((r) => `/${r.replace(/(\/?index)?\.js$/, '')}`),
  // sort wildcards last
  R.sortBy((r) => [
    100 + r.split('[[...').length,
    100 + r.split('[...').length,
    100 + r.split('[').length,
    999 - r.split('/').length,
    r,
  ].join('-')),
  // turn brackets into regex map
  R.reduce((acc, r) => ({
    ...acc,
    [r]: `^${r}$`
      .replace(/\/\[\[\.{3}[\w-]+]]/g, '(\\/.*)?')
      .replace(/\[\.{3}[\w-]+]/g, '(.+)')
      .replace(/\[[\w-]+]/g, '([^\\/]+)'),
  }), {}),
)();
```

`./LinkIn.js`: This component wraps next/link, and is itself wrapped by `Link`, which handles external links, mailto links, and downloads. 

```js
import NextLink from 'next/link';
import pt from 'prop-types';
import R from 'ramda';
import React from 'react';

import rawRoutes from /* preval */ './prevalRoutes';

// turn regex strings into RegExps
const routes = R.map((reg) => new RegExp(reg), rawRoutes);

const getHref = (to) => (
  routes[to] ? to :
  // take the first match and exit quick
  R.reduce((pick, route) => (
    routes[route].test(to) ? R.reduced(route) : null
  ), null, R.keys(routes))
);

export function LinkIn(props) {
  const {prefetch, replace, scroll, shallow, to, ...rest} = props;
  const linkProps = {
    as: to,
    href: getHref(to),
    prefetch,
    scroll,
    shallow,
    replace,
  };

  return (
    <NextLink {...linkProps} passHref>
      { /* can improve by handling children the NextLink way. */ }
      <a {...rest} />
    </NextLink>
  );
}
```

**Comment by mbrowne** [comment]
It would be ideal to make this as scalable as possible, so there's no additional run-time cost even in the case of thousands of route patterns. The `href` prop for each link could be generated at build-time, so something like the following except done internally by next.js (since it would be annoying to do this manually for every link—kind of the whole goal to avoid such manual work):

```js
  <Link as="/foo/1" href={preval.require('utils/getHref')("/foo/1")} >
    <a>Click me</a>
  </Link>
```

**Comment by martpie** [comment]
Well, I am on the opposite spectrum, I want things to be was dynamic as possible because I may have links coming from pages content from a headless CMS.

I guess we all have our usecase ;)

**Comment by mbrowne** [comment]
@martpie Well, the `href` could be generated at run-time instead of build-time in the case of SSR; I was more concerned about the impact on the client-side and static pages. Would it not be possible to have the best of both worlds? Or would you need it to be dynamic on the client as well? I suppose there could be a configuration option for those who didn't want to follow the standard convention (whatever the default ends up being).

**Comment by prowebat** [comment]
Dear All,

i have written a wrapper for next/link to accept href and any additional query parameters:
eg.
`<Link href="/news/[id]" query={{ id: 1234, anotherQueryParam: 567 }}>`
results to:
`<Link href="/news/[id]?anotherQueryParam=567" as="/news/1234?anotherQueryParam=567">`

i think that is a quite easy solution for the initials posters proposal.
code is as follows:
```js
import NextLink from "next/link";

export const Link = ({ query, ...props }) => {
  if (props.href.match && props.href.match(/\[/) && !props.as && query) {
    // encode placeholders in href string eg. /posts/[id], /posts/[...params]
    // and remove the placeholder from the query parameter
    props.as = props.href.replace(/\[([^\]]+)\]/g, (match, contents, offset, input_string) => {
      let paramName = contents;

      let ret;

      // test for [...params]
      if (paramName.match(/^\./)) {
        paramName = paramName.replace(/^\.+/, '');
        ret = query[paramName].join('/');
      } else {
        ret = query[paramName];
      }

      // remove from object:
      query = { ...query };
      delete query[paramName];

      return ret;
    });
  }

  if (Object.values(query).length) {
    // append query string to as and href (needed by nextjs)
    if (props.as) {
      props.as += "?" + toQueryString(query);
    }
    if (props.href) {
      props.href += "?" + toQueryString(query);
    }
  }
  return <NextLink {...props} />;
};

function toQueryString(params) {
// some implementation copied from stackoverflow...
}

```

regards.

**Comment by simka** [comment]
@Timer - I noticed you've added new label recently, does this mean it is something you are getting closer to implementing?

We could really use this feature in a large app we're migrating to Next.js currently. Have you determined or at least have an idea on how you want to solve it? Maybe it's something we could help with, instead of spending time on adjusting props for all the links in the app or patching it on our own and risking breakage in the future?

**Comment by flybayer** [comment]
I'm really interested in this too. We have a [similar Blitz RFC](https://github.com/blitz-js/blitz/discussions/585) on this.

The other big thing missing from the current Next link is `activeClassName` and `activeStyle` for styling links that match the current route.

**Comment by JonathanReisCom** [comment]
I solved this problem like this:

next.config.js:
```
const assetPrefixForNamespace = (namespace) => {
  switch (namespace) {
    case 'dev':
      return {
        url: '',
        exportTrailingSlash: false,
      };
    case 'ghpages':
      return {
        url: '/jonathanreis-website', // My GitHub Pages folder
        exportTrailingSlash: true,
      };
  }
};
const config = assetPrefixForNamespace(process.env.NAMESPACE);

module.exports = withPlugins([[withSass], [withImages]], {
  env: {
    APP_URL: config.url,
    NEED_DIRECT_LINK: config.exportTrailingSlash,
  },

  // Folders with index.html on export
  exportTrailingSlash: config.exportTrailingSlash,

  // Set prefix for absolute links
  assetPrefix: config.url,

  webpack: (config) => {
    config.output.publicPath = `${config.url}${config.output.publicPath}`;
    config.resolve.modules.push(path.resolve('./'));
    return config;
  },
});
```
Scripts for package.json:
```
"scripts": {
  "dev": "NAMESPACE=dev next",
  "start": "NAMESPACE=dev next start",
  "----": "echo '...'",
  "deploy_ghpages": "yarn l_a && yarn l_b && yarn pd_c && yarn pd_d && yarn pd_e && yarn pd_f && yarn pd_g",
  "l_a": "echo 'Cleaning Babel...' && rm -rf node_modules/.cache/babel-loader || echo 'Nothing to remove.' ",
  "l_b": "echo 'Installing dependencies...' && yarn install",
  "pd_c": "echo 'Remove old build...' && rm -r out/* || echo 'Nothing to remove.' ",
  "pd_d": "echo 'Building...' && NAMESPACE=ghpages next build",
  "pd_e": "echo 'Exporting...' && NAMESPACE=ghpages next export -o out/",
  "pd_f": "echo 'Copy .nojekyll...' && cp .nojekyll out/",
  "pd_g": "echo 'Deploying...' && gh-pages -t true -d out/"
},
```

Create a link:
```
<CustomLink href="/[language]/heroes-villains" as="/pt-br/heroes-villains">
  <a>Go to page</a>
</CustomLink>
```

CustonLink.js:
```
import React from 'react';
import Link from 'next/link';

const CustomLink = ({ children, href, as }) => {
  if (process.env.NEED_DIRECT_LINK) {
    return React.cloneElement(children, { href: `${process.env.APP_URL}${as}` });
  }

  return (
    <Link href={href} as={as}>
      {children}
    </Link>
  );
};

export default CustomLink;
```

If it's useful to anyone, let me know.

**Comment by remidej** [comment]
Just want to add my +1 to this issue because Next's links handling makes things quite hard for my use case.

I'm using Next.js to build static pages based on the contents of a headless CMS. I have no idea what kind of links (external, static route, dynamic routes) the CMS users will include.

So I just render a plain `<a>` tag in most cases, but I wish I could benefit from Next's frontend routing.

And I don't know if it's possible, but I think the best API would be to get rid of the Link component altogether, and to have a smarter behavior for `<a>` tags

**Comment by flybayer** [comment]
[Here's a nice custom `<Link>` component](https://gist.github.com/flybayer/f9ca161efd3abcb1d8cb76c92e05f5d2) I started using on a project. It can be used like this:

```tsx 
<Link page="/users/[id]" params={{ id: user.id }}>
  {user.name}
</Link>

<Link page="/blog/[...slug]" params={{ slug: ['coffee', 'frenchpress'] }}>
  View Here
</Link>
```

It's a a lot nicer `<Link>` to use until this RFC gets adopted (hopefully)

**Comment by jamesknelson** [comment]
It'd be great to have some way to use this feature programmatically as well. For example, with a new function like `Router.parse(url)`  that maps a URL to its appropriate filesystem route string and params. You could also return a Promise to if you want to do this on the server (and avoid sending a route manifest).

```js
await Router.parse("/news/1234") // { pathname: "/news/[id]", params: { id: "1234" } }
```

A function like this would come in handy in a bunch of situations. In particular, it'd make it a lot easier to move an existing react-router based CRA app into Next.js.

Re: `<Link to>` itself, there's no need to know the Next.js route name until the user clicks the link, so I think it makes to do the actual parsing lazily. On desktop, you could do it on mouseenter over the link, although mobile is trickier. Perhaps lazily parse links that are visible on the screen?

**Comment by thesunny** [comment]
I don't have time to package it up nicely, but I've been using this since dynamic routes came out and it's worked for everything I've needed.

The way I create links and do navigation is I created a `Go` object that uses `segments` to express dynamic URLS.

```js
<Go.Link segments={["news", {id: 1234}]}>
  <a>Link text</a>
</Go.Link>

And for pushing routes, I use:

Go.push(["news", {id: 1234}])
```

Basically, a string gets converted directly to part of the path. An object converts its "key" to the `href` and the "value" to the `as`.

Here's the code in TypeScript. May need to be cleaned but in case it's helpful, thought I'd post:

```ts
import kindOf from "kind-of"
import NextLink from "next/link"
import Router from "next/router"
import React from "react"
import invariant from "tiny-invariant"

type AtLeastOne<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> & U[keyof U]

/**
 * Note:
 * Lots of valuable information in `next/router`
 * <https://github.com/zeit/next.js/blob/canary/packages/next/client/router.ts>
 *
 * properties
 * `['pathname', 'route', 'query', 'asPath', 'components']`
 * methods
 * `['push', 'replace', 'reload', 'back', 'prefetch', 'beforePopState']`
 * events
 * `['routeChangeStart', 'beforeHistoryChange', 'routeChangeComplete', 'routeChangeError', 'hashChangeStart', 'hashChangeComplete']`
 */

type StringObject = AtLeastOne<{
  [key: string]: string
}>

export type Segment = string | StringObject

const Go = {
  /**
   * Takes an array of Object, parses them and return an `href` and an `as`
   */
  parse(segments: Segment[]) {
    const hrefSegments: string[] = []
    const asSegments: string[] = []
    segments.forEach((segment) => {
      switch (kindOf(segment)) {
        case "string":
          hrefSegments.push(`/${segment}`)
          asSegments.push(`/${segment}`)
          break
        case "object":
          Object.entries(segment).forEach(
            ([key, value]: [string, string | undefined]) => {
              if (value === undefined) throw new Error(`Should not happen`)
              hrefSegments.push(`/[${key}]`)
              asSegments.push(`/${value}`)
            }
          )
          break
        default:
          invariant(
            false,
            `Segments mut be a string or object but is ${segment}`
          )
      }
    })
    let href = hrefSegments.join("")
    let as = asSegments.join("")
    if (href === "") href = "/"
    if (as === "") as = "/"
    return { href, as }
  },
  /**
   * Navigate to the given URL like a `Router.push` but parses the segments
   */
  push(segments: Segment[], query?: StringObject) {
    const { href, as } = Go.parse(segments)
    Router.push({ pathname: href, query }, { pathname: as, query })
  },
  /**
   * Navigate to the given URL like a `Router.replace` but parses the segments
   */
  replace(segments: Segment[], query?: StringObject) {
    const { href, as } = Go.parse(segments)
    Router.replace({ pathname: href, query }, { pathname: as, query })
  },
  /**
   * Refresh page in browser.
   *
   * You can use `Router.reload` to reload on the server
   */
  refresh() {
    Router.replace(Router.pathname, Router.asPath)
  },
  /**
   * Reload page on server.
   */
  reload() {
    Router.reload()
  },
  /**
   * Returns a function that can be used as an event handler.
   */
  handler(segments: Segment[]) {
    return function (e: React.SyntheticEvent) {
      e.preventDefault()
      e.stopPropagation()
      Go.push(segments)
    }
  },
  Link: React.forwardRef(
    (
      {
        segments,
        query,
        ...props
      }: { segments: Segment[]; query: any; [key: string]: any },
      ref: React.Ref<any>
    ) => {
      const { href, as } = Go.parse(segments)
      return (
        <NextLink
          {...props}
          href={{ pathname: href, query }}
          as={{ pathname: as, query }}
          ref={ref}
        />
      )
    }
  ),
}

export { Go }
```

**Comment by fabb** [comment]
@mikestopcontinues cool solution!

Could it be that you mixed up the regexes for catch-all and regular dynamic path segments?

> ```js
>       .replace(/\[\.{3}[\w-]+]/g, '([^\\/]+)')
>       .replace(/\[[\w-]+]/g, '(.+)'),
> ```

The first line would map `/some/[...slugs]` to the regex `^/some/([^\\/]+)$` which would only match one path segment, but actually next.js uses this to match unlimited path segments.

The second line would map `/some/[slugs]` to the regex `^/some/(.+)$` which would match any number of path segments, but next.js would only match a single path segment.

Apart from this, the solution does not support [Optional catch all routes](https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes) yet, and it does not take into account that [predefined routes take precedence over dynamic routes](https://nextjs.org/docs/routing/dynamic-routes#caveats). Still a good temporary solution until next.js solves this on their end.

**Comment by nevnein** [comment]
Reading the docs of 9.5.3 I assume this is ready and shipped, is it so? Can we use `Link`s without `as`?

**Comment by mikestopcontinues** [comment]
@fabb Quite right! Thanks! 😄  I updated the code above to fix that bug as well as add support for optional catch-alls. I think the sort lines should ensure precedence. If not, let me know!

**Comment by jnv** [comment]
> Reading the docs of 9.5.3 I assume this is ready and shipped, is it so? Can we use `Link`s without `as`?

Not in stable version but per [this tweet](https://twitter.com/timneutkens/status/1295355450846846976?s=20) it's available in canary.

**Comment by balazsorban44** [comment]
@jnv, @nevnein Using it in `9.5.3` (not canary), without `as`. Works just fine! 🎉

**Comment by rokinsky** [comment]
> @jnv, @nevnein Using it in `9.5.3` (not canary), without `as`. Works just fine! 🎉

Cool. This RFC issue can be finally closed. Am I right, @Timer?

**Comment by fabb** [comment]
Wait a second before closing! We don‘t use `Link`, but but have a custom implementation based on `useRouter`. Is there now a way with public api to get the correct `href` from the `as` path?

**Comment by fabb** [comment]
Ok I just found this in the docs:

> `url` - The URL to navigate to
> `as` - Optional decorator for the URL that will be shown in the browser. Before Next.js 9.5.3 this was used for dynamic routes, check our previous docs to see how it worked
> https://nextjs.org/docs/api-reference/next/router#routerpush

So when we want to navigate to a path `/some/url` Where we have a dynamic next route `some/[path]`, we can now just pass `/some/url` as `url` and leave `as` empty, correct? That‘s awesome!

**Comment by timneutkens** [comment]
> we can now just pass /some/url as url and leave as empty, correct? That‘s awesome!

Correct!

> Cool. This RFC issue can be finally closed.

We've only covered the href resolving case (still needs to be announced publicly). That part is fully backwards compatible. Removing the need of `<a>` is also part of this RFC.

**Reply by realw5** [reply]
How are dynamic routes such as /category/[...slug] handled when a rewrite is used? Prefetch seems to not work in this case. 

For example:

Page Route:  /pages/category/[...slug].tsx
Rewrite Rule: /moveis/moveis-para-escritorio/mesas-para-computador -> /category/moveis/moveis-para-escritorio/mesas-para-computador

Everything works just fine except for prefetching. I have tried:
 <Link href="/moveis/moveis-para-escritorio/mesas-para-computador"> - No Prefetching
 <Link href="/category/[...slug]" as="/moveis/moveis-para-escritorio/mesas-para-computador"> - Broken Prefetching (seems to be fetching the current page, not the the "as" link

Am I missing something?

**Comment by fabb** [comment]
It looks like now it's possible to have trailing slashes in urls when pushing. Nice!

**Comment by equinusocio** [comment]
Came here from https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-is-valid.md.

Cannot see a way to use nextjs Link with the eslint rule `jsx-a11y/anchor-is-valid` without disabling it every time, cause missing href on `a` elements

**Comment by timneutkens** [comment]
This RFC is now obsolete as the individual features have been landed in a slightly different way. Resolving the `href` automatically has been available for some time and `<a>` will be no longer required when https://github.com/vercel/next.js/pull/36436 becomes available on stable in Next.js 13.
